frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java

```diff
@@ -344,18 +347,42 @@ private ContainerRequest setupContainerRequest(TaskStatus taskStatus) throws Exc
           requestResource, maxResource);
     }
 
-    if (requestResource.getGpuNumber() > 0) {
-      updateNodeReports(yarnClient.getNodeReports(NodeState.RUNNING));
+    ResourceDescriptor optimizedRequestResource = YamlUtils.deepCopy(requestResource, ResourceDescriptor.class);
+
+    SelectionResult selectionResult = null;
 
-      SelectionResult selectionResult = selectionManager.select(requestResource, requestNodeLabel, requestNodeGpuType);
-      if (selectionResult != null) {
-        ResourceDescriptor optimizedRequestResource = YamlUtils.deepCopy(requestResource, ResourceDescriptor.class);
-        optimizedRequestResource.setGpuAttribute(selectionResult.getGpuAttribute());
-        return HadoopUtils.toContainerRequest(optimizedRequestResource, requestPriority, null, selectionResult.getNodeHost());
+    //user request port number is bigger than user specific port count, dynamic allocate the vacancy number of port.
+    int newRequestNumber = optimizedRequestResource.getPortNumber() - PortRangeUtils.getPortsNumber(optimizedRequestResource.getPortRanges());
```

Comment:
>getPortNumber [](start = 52, length = 13)

PortNumber means the number of ports or the value of a port?

Reply:
portNumber means the number of ports,  previously we use gguNumber to indicate GPU count, to keep the same style in ResourceDescriptor, here we use portNumber instead of portCount. 

---
In reply to: [166912573](https://github.com/Microsoft/pai/pull/318#discussion_r166912573) [](ancestors = 166912573)